File: Jump_to_Algorithm2/week31/dayeongYoo/괄호 회전하기.java
// 여러가지 괄호가 섞인 문자열을 입력값
// 이 괄호를  문자열의 길이만큼 회전시키고,  회전시킨값이 '올바른 괄호 문자열'이라면,
// 결과값을 +1 늘리기

import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        String str = s;

        for (int i = 0; i < s.length(); i++) {
            char temp = str.charAt(0);
            str = str.substring(1) + temp;
            if (isCorrect(str)) {
                answer++;
            }
        }

        return answer;
    }

    public boolean isCorrect(String str) {
        Stack<Character> s = new Stack<>();

        char[] arr = str.toCharArray();

        for (char data : arr) {
            if (s.isEmpty()) {
                s.push(data);
            } else if (s.peek() == '[' && data == ']') {
                s.pop();
            } else if (s.peek() == '(' && data == ')') {
                s.pop();
            } else if (s.peek() == '{' && data == '}') {
                s.pop();
            } else {
                s.push(data);
            }
        }

        if (s.isEmpty()) {
            return true;
        }

        return false;
    }
}

File: Jump_to_Algorithm2/week31/KyoungyeonJu/괄호 회전하기.py
def solution(s):
    answer = 0
    l = len(s)
    arr = list(s)

    for i in range(l):
        narr = []  # 스택
        Flag = True  # 올바른 괄호 문자열인지 확인
        for j in arr:
            if j == '(' or j == '[' or j == '{':  # 열리는 괄호
                narr.append(j)
            else:  # 닫히는 괄호
                if len(narr) == 0:  # 스택 empty인 경우, False
                    Flag = False

                elif j == ')':
                    if narr.pop() == '(': continue
                    else: Flag = False
                elif j == '}':
                    if narr.pop() == '{': continue
                    else: Flag = False
                elif j == ']':
                    if narr.pop() == '[': continue
                    else: Flag = False

            if Flag is False:
                break

        if Flag and not narr:  # 올바른 괄호 문자열이면서 빈 스택이면, 테케13 "(){{"
            answer += 1

        # 문자열 회전
        front = arr.pop(0)
        arr.append(front)

    return answer


"""
"[](){}"	3
"}]()[{"	2
"[)(]"	0
"}}}"	0
"(){{"
"""